File: project_pyratio/databases.py
# databases.py
# Database type for reading, storing, and supplying all forms of DSP game data

from facilities import FacilityCategory, Facility
from products import Product
from rates import Rate, Time
from recipes import Recipe


#######################################
# DataBase Type and Inner Constructor #
#######################################

class DataBase:

    def __init__(self) -> None:
        self.facility_categories: set[FacilityCategory] = set()
        self.facilities: dict[str, Facility] = {}
        self.products: set[Product] = set()
        self.recipes: dict[str, Recipe] = {}
        self.rates: dict[str, Rate] = {}
        self.times: dict[str, Time] = {}
        self.lookup_table_facilities: dict[FacilityCategory, set[Facility]] = {}
        self.lookup_table_recipes: dict[Product, set[Recipe]] = {}

    ###############################
    # Lookup Table Initialization #
    ###############################

    def _make_table_facilities(self) -> 'DataBase':
        for facility in self.facilities.values():
            if facility.category not in self.lookup_table_facilities:
                self.lookup_table_facilities[facility.category] = set()
            self.lookup_table_facilities[facility.category].add(facility)
        return self

    def _make_table_recipes(self) -> 'DataBase':
        for recipe in self.recipes.values():
            for quantity in recipe.outputs:
                if quantity.product not in self.lookup_table_recipes:
                    self.lookup_table_recipes[quantity.product] = set()
                self.lookup_table_recipes[quantity.product].add(recipe)
        return self

    def make_tables(self) -> 'DataBase':
        print("Making lookup tables for DataBase...", end="", flush=True)
        self._make_table_facilities()
        self._make_table_recipes()
        print(" done!")
        return self

    ############################
    # DataBase Query Functions #
    ############################

    def find_facilities(
        self,
        category: FacilityCategory | str | Recipe
    ) -> set[Facility]:
        if isinstance(category, str):
            category = FacilityCategory(category)
        elif isinstance(category, Recipe):
            category = category.category
        return self.lookup_table_facilities.get(category, set())

    def find_recipes(
        self,
        recipe_output: Product | str
    ) -> set[Recipe]:
        if not isinstance(recipe_output, Product):
            recipe_output = Product(recipe_output)
        return self.lookup_table_recipes.get(recipe_output, set())

File: julia_code/databases.jl
# databases.jl
# Database type for reading, storing, and supplying all forms of DSP game data


#######################################
# DataBase Type and Inner Constructor #
#######################################

struct DataBase
	
	facilitycategories::Set{FacilityCategory}
	facilities::Dict{String, Facility}
	
	products::Set{Product}
	
	recipes::Dict{String, Recipe}
	
	rates::Dict{String, Rate}
	times::Dict{String, Time}
	
	lookuptable_facilities::Dict{FacilityCategory, Set{Facility}}
	lookuptable_recipes::Dict{Product, Set{Recipe}}
	
	function DataBase()
		# This is either really clever or really bad practice... not sure...
		# I just want to concisely initialize with every data structure empty
		new((FieldType() for FieldType in fieldtypes(DataBase))...)
	end
end


###############################
# Lookup Table Initialization #
###############################

function _maketable_facilities(db::DataBase)
	
	for facility in values(db.facilities)
		if !haskey(db.lookuptable_facilities, category(facility))
			db.lookuptable_facilities[category(facility)] = Set{Facility}()
		end
		push!(db.lookuptable_facilities[category(facility)], facility)
	end
	
	db
end

function _maketable_recipes(db::DataBase)
	
	for recipe in values(db.recipes)
		for quantity in outputs(recipe)
			if !haskey(db.lookuptable_recipes, product(quantity))
				db.lookuptable_recipes[product(quantity)] = Set{Recipe}()
			end
			push!(db.lookuptable_recipes[product(quantity)], recipe)
		end
	end
	
	db
end

function maketables!(db::DataBase)
	
	print("Making lookup tables for DataBase...")
	flush(stdout)
	
	_maketable_facilities(db)
	_maketable_recipes(db)
	
	println(" done!")
	db
end


############################
# DataBase Query Functions #
############################

function findfacilities(db::DataBase, category::FacilityCategory)
	get!(Set{Facility}, db.lookuptable_facilities, category)
end

function findfacilities(db::DataBase, categoryname::AbstractString)
	findfacilities(db, FacilityCategory(categoryname))
end

function findfacilities(db::DataBase, recipe::Recipe)
	findfacilities(db, category(recipe))
end

function findrecipes(db::DataBase, recipeoutput::Product)
	get!(Set{Recipe}, db.lookuptable_recipes, recipeoutput)
end

function findrecipes(db::DataBase, recipeoutputname::AbstractString)
	findrecipes(db, Product(recipeoutputname))
end

